pp_ctl.c

```diff
@@ -4284,40 +4319,113 @@ S_require_file(pTHX_ SV *sv)
                         SvGETMAGIC(loader);
                     }
 
-                    Perl_sv_setpvf(aTHX_ namesv, "/loader/0x%" UVxf "/%s",
-                                   PTR2UV(SvRV(dirsv)), name);
-                    tryname = SvPVX_const(namesv);
-                    tryrsfp = NULL;
-
                     if (SvPADTMP(nsv)) {
                         nsv = sv_newmortal();
                         SvSetSV_nosteal(nsv,sv);
                     }
 
-                    ENTER_with_name("call_INC");
-                    SAVETMPS;
-                    EXTEND(SP, 2);
+                    const char *method = NULL;
+                    bool is_incdir = FALSE;
+                    SV * inc_idx_sv = save_scalar(PL_incgv);
+                    sv_setiv(inc_idx_sv,inc_idx);
+                    if (sv_isobject(loader)) {
+                        /* if it is an object and it has an INC method, then
+                         * call the method. Otherwise assume its a blessed coderef.
+                         */
+                        HV *pkg = SvSTASH(SvRV(loader));
+                        GV * gv = gv_fetchmethod_pvn_flags(pkg, "INC", 3, 0);
+                        if (gv && isGV(gv)) {
+                            method = "INC";
+                        } else {
+                            gv = gv_fetchmethod_pvn_flags(pkg, "INCDIR", 6, 0);
+                            if (gv && isGV(gv)) {
+                                method = "INCDIR";
+                                is_incdir = TRUE;
+                            }
+                        }
+                        if (!method) {
+                            if (SvTYPE(SvRV(loader)) != SVt_PVCV) {
```

Comment:
Should this allow `&{}` overloads?

Reply:
Hrm. I dunno. On one hand, yeah from a completeness point of view we probably should. On the other hand, we don't support objects with overloaded `@{}` to implement the array API either, and in fact historically we would assume any object in @INC would have an INC method, and blindly call it as a method, with the resulting errors when it was a blessed sub with no methods. Its a tricky call. Prior to this patch you couldn't put an object in the array anyway, so nothing is going to break...

Reply:
If we only allow unblessed array refs, it feels like we should only allow unblessed code refs.

Reply:
That doesn't make sense. If the array ref is blessed it is treated as an object, not as a hook wrapper. If it is unblessed then we treat it as a hook wrapper.  That is existing logic and can't change.

We used to ONLY allow subrefs in the zero slot of the wrapper, which made some sense because the only hook method we offered was INC. However we now have two methods, INCDIR and INC.  If we wish to have the wrapper approach work with both hooks we need to be able to supply an object in the zero slot.

The thread here doesnt show the full context.

```
                    if (sv_isobject(loader)) {
                        /* if it is an object and it has an INC method, then
                         * call the method. Otherwise assume its a blessed coderef.
                         */
                        HV *pkg = SvSTASH(SvRV(loader));
                        GV * gv = gv_fetchmethod_pvn_flags(pkg, "INC", 3, 0);
                        if (gv && isGV(gv)) {
                            method = "INC";
                        } else {
                            gv = gv_fetchmethod_pvn_flags(pkg, "INCDIR", 6, 0);
                            if (gv && isGV(gv)) {
                                method = "INCDIR";
                                is_incdir = TRUE;
                            }
                        }
                        if (!method) {
                            if (SvTYPE(SvRV(loader)) != SVt_PVCV) {
                                if (dirsv != loader)
                                    croak("Object with arguments in @INC does not support a hook method");
                                else
                                    goto treat_as_string;
                            }
                        }
                    }
```

So, if the loader is an object (either wrapper or unwrapped), we check to see if it has either hook methods. If it does not then we check if it is a blessed sub, and if it is then we call it, if it is not then we treat the object as a string.  In the special case where we have a wrapper (`dirsv != loader`), we are bit more careful, it doesn't make a lot of sense to use the wrapper api and include a non-coderef, non-hook bearing object. Unlike the unwrapped case, it doesn't make any sense to ignore the arguments (the wrapper), and fallback to strignifying the ref. Why would you provide a wrapper if it is just going to be thrown away to be stringified? 

In the old code, this would have worked differently. We would have seen the item was an object, tried to call the INC method and died with a cryptic error message.

With the new implementation, we expand things so that this:

```
push @INC, bless(sub { warn "@INC" }, "Me");
```

work the same as

```
push @INC, sub { warn "@INC" };
```

I believe that this is no bad thing. It helps debugging `@INC` hooks, and it doesn't break anything existing, and IMO it is more naturally DWIM anyway, the rules can say "An object with an INC or INCDIR hook, or a coderef". That isn't inconsistent with a blessed coderef with no method.